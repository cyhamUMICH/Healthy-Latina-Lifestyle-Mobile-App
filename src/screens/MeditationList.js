import React, { useEffect, useState } from 'react';
import { View } from 'react-native';
import { styles } from '../styles/Styles';
import ContentList from '../components/ContentList';
import LoadingSpinner from '../components/LoadingSpinner';
import * as firebase from 'firebase/app';
import "firebase/firestore";
import "firebase/storage";

const MeditationList = (props) => {
  const [data, setData] = useState([]);           
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    const fetchList = async () => {
      const dbh = firebase.firestore();
      dbh.collection("meditations").get()
      .then((querySnapshot) => {

        if (querySnapshot.size == 0) {
          setIsLoaded(true);
        }
        else {
          let countImages = 0;
          querySnapshot.forEach((doc) => {
            let newDoc = doc.data();
            newDoc.contentID = doc.id;
            
            // https://firebase.google.com/docs/storage/web/download-files
            let storage = firebase.storage();
            let pathReference = storage.ref(newDoc.imagePath);
            pathReference.getDownloadURL()
            .then((url) => {
              newDoc.imagePath = url;
            })
            .catch((error) => {
              newDoc.imagePath = "";
            })
            .finally(() => {
              // Add to the data list once the image has been resolved
              setData(oldList => [...oldList, newDoc]);
              
              countImages++;

              // Data is loaded once the number of images is the same as
              // the number in the snapshot
              if (querySnapshot.size === countImages) {
                setIsLoaded(true);
              }
            });
          })
        }
      })
      .catch((error) => {
        console.log("Error getting documents: ", error);
      });
    };
  
    fetchList();  
  }, []);

  return (
  
    <View style={styles.app}>
    {
      isLoaded ? 
        <ContentList 
          contentComponent="Meditation"
          navigation={props.navigation}
          contentType="Meditations"
          data={data.sort((docA, docB) => docB.dateAdded - docA.dateAdded)}
          filterBy="Difficulty,Language,Topic,Duration" />
        : <LoadingSpinner />
    }
    </View>
  );
};

export default MeditationList;