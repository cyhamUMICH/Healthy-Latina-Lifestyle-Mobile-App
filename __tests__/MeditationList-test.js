// Code from: https://github.com/byCedric/expo-guide-ci/blob/master/__tests__/App-test.js
//      and https://jestjs.io/docs/tutorial-react-native
import React from 'react';
import MeditationList from '../src/screens/MeditationList';
import renderer from 'react-test-renderer';

test('renders the loading screen', async () => {
  const tree = renderer.create(<MeditationList />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders the root without loading screen', async () => {
  const tree = renderer.create(<MeditationList skipLoadingScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});