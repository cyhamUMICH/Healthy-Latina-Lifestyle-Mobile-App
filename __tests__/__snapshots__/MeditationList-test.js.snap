// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Meditation List With the Loading Screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#00A89D"
      hidesWhenStopped={true}
      size="large"
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 2.5,
            },
          ],
        }
      }
    />
  </View>
</View>
`;

exports[`Renders Meditation List Without the Loading Screen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#00A89D"
      hidesWhenStopped={true}
      size="large"
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 2.5,
            },
          ],
        }
      }
    />
  </View>
</View>
`;
